File: old/source/frontend/typecheck.cpp
#include <config.hpp>

#include <iostream>

static ExpressionKind primitive_as_kind[] =
{
    ExpressionKind::Void,
    ExpressionKind::Bool,
    ExpressionKind::String,

    ExpressionKind::Integer,
    ExpressionKind::Integer,
    ExpressionKind::Integer,
    ExpressionKind::Integer,
    ExpressionKind::Integer,
    ExpressionKind::Integer,
    ExpressionKind::Integer,
    ExpressionKind::Integer,

    ExpressionKind::Floating,
    ExpressionKind::Floating,
};

// Compare types
static bool compare(ExpressionKind destination, ExpressionKind source)
{
    if (source == ExpressionKind::Integer && destination == ExpressionKind::Floating)
        return true;

    return (source == destination);
}

// Type checking
ExpressionKind Node::typecheck(Input *input)
{
    return ExpressionKind::Void;
}

// Type checking
ExpressionKind NodeBinary::typecheck(Input *input)
{
    return ExpressionKind::Void;
}

// Type checking
ExpressionKind NodeUnary::typecheck(Input *input)
{
    return ExpressionKind::Void;
}

// Type checking
ExpressionKind NodeNumber::typecheck(Input *input)
{
    if ((static_cast<int64>(value) - value) == 0)
        return ExpressionKind::Integer;
    else
        return ExpressionKind::Floating;
}

// Type checking
ExpressionKind NodePrimitive::typecheck(Input *input)
{
    return primitive_as_kind[(int32)kind];
}

// Type checking
ExpressionKind NodeImport::typecheck(Input *input)
{
    return ExpressionKind::Void;
}

// Type checking
ExpressionKind NodeFunction::typecheck(Input *input)
{
    input->current_function = this;
    block->typecheck(input);

    return ExpressionKind::Void;
}

// Type checking
ExpressionKind NodeParameter::typecheck(Input *input)
{
    return type->typecheck(input);
}

// Type checking
ExpressionKind NodeBlock::typecheck(Input *input)
{
    for (Node *statement : statements)
        statement->typecheck(input);

    return ExpressionKind::Void;
}

// Type checking
ExpressionKind NodeReturn::typecheck(Input *input)
{
    ExpressionKind type          = ((!value) ? ExpressionKind::Void : value->typecheck(input));
    ExpressionKind function_type = input->current_function->type->typecheck(input);
    
    if (!compare(function_type, type))
        input->logger.error(start, "Return type does not match function return type.");

    return ExpressionKind::Void;
}

File: source/frontend/passes/TypeCheck.cs
using System;

namespace Suteki
{
    partial class NodePrimitive : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            switch (Kind)
            {
                case PrimitiveKind.Void:
                    return ExpressionKind.Void;

                case PrimitiveKind.Bool:
                    return ExpressionKind.Bool;

                case PrimitiveKind.String:
                    return ExpressionKind.String;

                case PrimitiveKind.Single:
                case PrimitiveKind.Double:
                    return ExpressionKind.Float;

                default:
                    return ExpressionKind.Integer;
            }
        }
    }

    partial class NodeBool : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            return ExpressionKind.Bool;
        }
    }

    partial class NodeInteger : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            return ExpressionKind.Integer;
        }
    }

    partial class NodeFloat : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            return ExpressionKind.Float;
        }
    }

    partial class NodeString : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            return ExpressionKind.String;
        }
    }

    partial class NodeFunction : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            if (Property != PropertyKind.Extern)
            {
                input.CurrentFunction = this;
                Block.TypeCheck(input);
            }
            
            return ExpressionKind.Void;
        }
    }

    partial class NodeParameter : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            return Type.TypeCheck(input);
        }
    }

    partial class NodeBlock : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            foreach (Node node in Statements)
                node.TypeCheck(input);

            return ExpressionKind.Void;
        }
    }

    partial class NodeCall : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            // TODO: make this shorter
            NodeFunction node = (NodeFunction)input.Globals[Name.Data.ToString()].Node;

            if (Parameters.Count != node.Parameters.Count)
            {
                input.Logger.Error(Name, "Function call parameter(s) count does not match function parameter(s) count.");
                return ExpressionKind.Void;
            }

            for (int index = 0; index < node.Parameters.Count; ++index)
            {
                ExpressionKind parameter  = node.Parameters[index].TypeCheck(input);
                ExpressionKind expression = Parameters[index].TypeCheck(input);

                if (!Type.Compare(parameter, expression))
                    input.Logger.Error(Name, $"Function call parameter ({index}) type does not match function parameter type.");
            }

            return ExpressionKind.Void;
        }
    }

    partial class NodeReturn : Node
    {
        public override ExpressionKind TypeCheck(Input input)
        {
            ExpressionKind expressionType = (Expression == null) ?
                                             ExpressionKind.Void :
                                             Expression.TypeCheck(input);
            ExpressionKind functionType   = input.CurrentFunction.Type.TypeCheck(input);

            // Compare types
            if (!Type.Compare(functionType, expressionType))
                input.Logger.Error(Start, "Return value type does not match function return type.");
    
            return ExpressionKind.Void;
        }
    }
}